File: exercicios/aula13/ex05_escopos/escopos.py
"""Exercício 05 — parser descendente com contexto herdado (escopos).

Amplia o exercício 04 com blocos aninhados. Gramática:

    P      -> item*                     (escopo "global")
    item   -> decl | bloco | uso
    decl   -> T id { , id } ;           T -> int | float | char | bool
    bloco  -> { item* }                 (abre um escopo filho do atual)
    uso    -> id = valor ;              valor -> id | num

Atributo herdado: o **escopo atual**. Ele desce como parâmetro explícito
(`parse_item(escopo)`, `parse_bloco(pai)`, ...) — não há variável global com
"o escopo corrente". Ao entrar num bloco, `Scope(pai)` cria o escopo filho;
ao sair (`}`), o parser simplesmente volta a usar o escopo que recebeu, e o
filho deixa de ser alcançável.

Regras semânticas:

- declarar um nome já declarado **no mesmo escopo** é erro (redeclaração);
- declarar um nome que existe num escopo ancestral é **sombreamento
  permitido**, registrado na AST (`sombreia x de global (1:5)`);
- cada uso procura o nome primeiro no escopo atual e depois nos ancestrais; o
  nó Id guarda a declaração resolvida (escopo, tipo e posição). Nome ausente é
  erro semântico na posição do uso.

A AST guarda linha e coluna de todos os nós, e os erros semânticos não
interrompem a análise (código de saída 4 no fim).

Uso:  python escopos.py <arquivo | ->
"""

import sys
from dataclasses import dataclass, field
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "comum"))
import fluxo as F  # noqa: E402

TIPOS = {"KW_INT": "int", "KW_FLOAT": "float", "KW_CHAR": "char", "KW_BOOL": "bool"}


# -- escopos ---------------------------------------------------------------


@dataclass
class Simbolo:
    nome: str
    tipo: str
    linha: int
    coluna: int
    escopo: str


class Scope:
    def __init__(self, nome: str, parent: "Scope | None"):
        self.nome = nome
        self.parent = parent
        self.symbols: dict[str, Simbolo] = {}

    def declare(self, simbolo: Simbolo):
        """Insere no escopo atual; devolve a declaração anterior do mesmo escopo, se houver."""
        anterior = self.symbols.get(simbolo.nome)
        if anterior is None:
            self.symbols[simbolo.nome] = simbolo
        return anterior

    def lookup(self, nome: str):
        """Procura no escopo atual e depois nos ancestrais."""
        escopo = self
        while escopo is not None:
            if nome in escopo.symbols:
                return escopo.symbols[nome]
            escopo = escopo.parent
        return None


# -- AST -------------------------------------------------------------------


@dataclass
class Id:
    nome: str
    linha: int
    coluna: int
    decl: Simbolo | None  # declaração resolvida (None = não declarado)


@dataclass
class Num:
    lexema: str
    linha: int
    coluna: int


@dataclass
class VarDecl:
    simbolo: Simbolo
    sombreia: Simbolo | None
    redeclaracao: bool


@dataclass
class Assign:
    alvo: Id
    valor: "Id | Num"
    linha: int
    coluna: int


@dataclass
class Block:
    escopo: str
    itens: list = field(default_factory=list)


def imprime(no, nivel: int = 0):
    ind = "  " * nivel
    match no:
        case Block(escopo=escopo, itens=itens):
            print(f"{ind}Block escopo={escopo}")
            for item in itens:
                imprime(item, nivel + 1)
        case VarDecl(simbolo=s, sombreia=sombra, redeclaracao=redecl):
            texto = f"{ind}VarDecl {s.nome} : {s.tipo} @{s.linha}:{s.coluna} [{s.escopo}]"
            if sombra is not None:
                texto += f" sombreia {sombra.nome} de {sombra.escopo} ({sombra.linha}:{sombra.coluna})"
            if redecl:
                texto += " redeclaração"
            print(texto)
        case Assign(alvo=alvo, valor=valor, linha=l, coluna=c):
            print(f"{ind}Assign @{l}:{c}")
            imprime(alvo, nivel + 1)
            imprime(valor, nivel + 1)
        case Id(nome=nome, linha=l, coluna=c, decl=d):
            destino = (
                f"{d.tipo} [{d.escopo}, declarado em {d.linha}:{d.coluna}]" if d else "não declarado"
            )
            print(f"{ind}Id {nome} @{l}:{c} -> {destino}")
        case Num(lexema=lex, linha=l, coluna=c):
            print(f"{ind}Num {lex} @{l}:{c}")


# -- parser ----------------------------------------------------------------


class Parser:
    def __init__(self, fl: F.Fluxo):
        self.fl = fl
        self.erros: list[str] = []
        self.blocos = 0  # só para dar nome aos escopos (bloco1, bloco2, ...)

    def parse_P(self) -> Block:
        global_ = Scope("global", None)
        raiz = Block("global")
        while not self.fl.verifica("EOF"):
            raiz.itens.extend(self.parse_item(global_))
        return raiz

    def parse_item(self, escopo: Scope) -> list:
        if self.fl.atual().type in TIPOS:
            return self.parse_decl(escopo)
        if self.fl.verifica("LBRACE"):
            return [self.parse_bloco(escopo)]
        if self.fl.verifica("IDENT"):
            return [self.parse_uso(escopo)]
        raise F.ErroSintaxe(self.fl.atual(), "declaração, bloco ou atribuição")

    def parse_bloco(self, pai: Scope) -> Block:
        self.fl.exige("LBRACE", "'{'")
        self.blocos += 1
        escopo = Scope(f"bloco{self.blocos}", pai)  # enter_scope
        bloco = Block(escopo.nome)
        while not self.fl.verifica("RBRACE"):
            if self.fl.verifica("EOF"):
                raise F.ErroSintaxe(self.fl.atual(), "'}'")
            bloco.itens.extend(self.parse_item(escopo))
        self.fl.proximo()
        return bloco  # leave_scope: `escopo` sai de alcance aqui

    def parse_decl(self, escopo: Scope) -> list:
        tipo = TIPOS[self.fl.proximo().type]
        decls = [self.declara(self.fl.exige("IDENT", "identificador"), tipo, escopo)]
        while self.fl.aceita("COMMA"):
            decls.append(self.declara(self.fl.exige("IDENT", "identificador"), tipo, escopo))
        self.fl.exige("SEMICOLON", "',' ou ';'")
        return decls

    def declara(self, tok, tipo: str, escopo: Scope) -> VarDecl:
        simbolo = Simbolo(tok.lexeme, tipo, tok.line, tok.column, escopo.nome)
        sombra = escopo.parent.lookup(tok.lexeme) if escopo.parent else None
        anterior = escopo.declare(simbolo)
        if anterior is not None:
            self.erros.append(
                F.erro_semantico(
                    tok.line,
                    tok.column,
                    f"redeclaração de '{tok.lexeme}' no escopo {escopo.nome} "
                    f"(declarado na linha {anterior.linha}, coluna {anterior.coluna})",
                )
            )
            return VarDecl(simbolo, None, True)
        return VarDecl(simbolo, sombra, False)

    def parse_uso(self, escopo: Scope) -> Assign:
        alvo = self.resolve(self.fl.proximo(), escopo)
        self.fl.exige("ASSIGN", "'='")
        tok = self.fl.atual()
        if tok.type == "IDENT":
            valor = self.resolve(self.fl.proximo(), escopo)
        elif tok.type == "INT":
            self.fl.proximo()
            valor = Num(tok.lexeme, tok.line, tok.column)
        else:
            raise F.ErroSintaxe(tok, "identificador ou número")
        self.fl.exige("SEMICOLON", "';'")
        return Assign(alvo, valor, alvo.linha, alvo.coluna)

    def resolve(self, tok, escopo: Scope) -> Id:
        decl = escopo.lookup(tok.lexeme)
        if decl is None:
            self.erros.append(F.erro_semantico(tok.line, tok.column, f"identificador '{tok.lexeme}' não declarado"))
        return Id(tok.lexeme, tok.line, tok.column, decl)


def main(argv: list[str]) -> int:
    F.configurar_saida()
    if len(argv) != 2:
        print("uso: python escopos.py <arquivo | ->", file=sys.stderr)
        return F.ERRO_USO
    fl, codigo = F.abrir(argv[1])
    if fl is None:
        return codigo
    p = Parser(fl)
    try:
        arvore = p.parse_P()
    except F.ErroSintaxe as erro:
        print(erro, file=sys.stderr)
        return F.ERRO_SINTAXE
    imprime(arvore)
    for e in p.erros:
        print(e, file=sys.stderr)
    return F.ERRO_SEMANTICO if p.erros else F.OK


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: exercicios/aula13/ex05_escopos/escopos.c
/* Exercício 05 — parser descendente com contexto herdado (escopos).
 *
 * Mesma gramática, mesmas regras e mesma saída de escopos.py (ver a docstring
 * de lá). Em C:
 *
 * - `Scope` tem ponteiro para o pai e um vetor dinâmico de símbolos;
 *   enter_scope(pai) cria o filho e leave_scope(s) o destrói e devolve o pai;
 * - **posse**: o Scope é dono dos seus símbolos e do seu nome, e morre no
 *   leave_scope. Por isso a AST **não aponta** para símbolos: cada nó copia o
 *   que precisa da declaração resolvida (nome do escopo, tipo, posição). Assim
 *   a árvore continua válida depois que os escopos foram destruídos;
 * - o escopo atual é sempre um parâmetro das funções do parser (nada global);
 * - em erro de sintaxe, cada parse_bloco fecha o próprio escopo antes de
 *   propagar o erro, e o main libera a AST parcial.
 *
 * Uso:  ./escopos <arquivo | ->
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../comum/fluxo.h"

/* -- escopos ------------------------------------------------------------ */

typedef struct {
    char *nome;
    const char *tipo;
    int linha, coluna;
} Simbolo;

typedef struct Scope {
    char *nome;
    struct Scope *parent;
    Simbolo *v;
    size_t n, cap;
} Scope;

static Scope *enter_scope(Scope *parent, char *nome) {
    Scope *s = aloca(sizeof(Scope));
    *s = (Scope){nome, parent, NULL, 0, 0};
    return s;
}

static Scope *leave_scope(Scope *s) {
    Scope *pai = s->parent;
    for (size_t i = 0; i < s->n; i++) free(s->v[i].nome);
    free(s->v);
    free(s->nome);
    free(s);
    return pai;
}

static Simbolo *procura_local(Scope *s, const char *nome) {
    for (size_t i = 0; i < s->n; i++) {
        if (strcmp(s->v[i].nome, nome) == 0) return &s->v[i];
    }
    return NULL;
}

/* Procura no escopo atual e depois nos ancestrais; *onde recebe o escopo. */
static Simbolo *lookup(Scope *s, const char *nome, Scope **onde) {
    for (; s; s = s->parent) {
        Simbolo *sim = procura_local(s, nome);
        if (sim) {
            *onde = s;
            return sim;
        }
    }
    return NULL;
}

static void declare(Scope *s, const Token *t, const char *tipo) {
    if (s->n == s->cap) {
        s->cap = s->cap ? s->cap * 2 : 8;
        s->v = realoca(s->v, s->cap * sizeof(Simbolo));
    }
    s->v[s->n++] = (Simbolo){copia(t->lexeme), tipo, t->line, t->column};
}

/* -- AST ------------------------------------------------------------------ */

typedef enum { K_BLOCK, K_DECL, K_ASSIGN, K_ID, K_NUM } Kind;

typedef struct No {
    Kind k;
    int linha, coluna;
    char *nome; /* BLOCK: nome do escopo; DECL/ID: identificador; NUM: lexema */
    /* DECL: tipo/escopo da própria declaração. ID: da declaração resolvida
     * (escopo NULL = não declarado), com a posição dela em dl:dc. */
    const char *tipo;
    char *escopo;
    int dl, dc;
    /* DECL: declaração sombreada (sombra NULL = nenhuma) e redeclaração. */
    char *sombra;
    int sl, sc;
    int redecl;
    struct No **f;
    size_t n, cap;
} No;

static No *no_novo(Kind k, int linha, int coluna, char *nome) {
    No *n = aloca(sizeof(No));
    *n = (No){.k = k, .linha = linha, .coluna = coluna, .nome = nome};
    return n;
}

static void no_filho(No *pai, No *f) {
    if (pai->n == pai->cap) {
        pai->cap = pai->cap ? pai->cap * 2 : 4;
        pai->f = realoca(pai->f, pai->cap * sizeof(No *));
    }
    pai->f[pai->n++] = f;
}

static void no_free(No *n) {
    if (!n) return;
    for (size_t i = 0; i < n->n; i++) no_free(n->f[i]);
    free(n->f), free(n->nome), free(n->escopo), free(n->sombra), free(n);
}

static void imprime(const No *n, int nivel) {
    printf("%*s", 2 * nivel, "");
    switch (n->k) {
    case K_BLOCK:
        printf("Block escopo=%s\n", n->nome);
        for (size_t i = 0; i < n->n; i++) imprime(n->f[i], nivel + 1);
        break;
    case K_DECL:
        printf("VarDecl %s : %s @%d:%d [%s]", n->nome, n->tipo, n->linha, n->coluna, n->escopo);
        if (n->sombra) printf(" sombreia %s de %s (%d:%d)", n->nome, n->sombra, n->sl, n->sc);
        if (n->redecl) printf(" redeclaração");
        putchar('\n');
        break;
    case K_ASSIGN:
        printf("Assign @%d:%d\n", n->linha, n->coluna);
        imprime(n->f[0], nivel + 1);
        imprime(n->f[1], nivel + 1);
        break;
    case K_ID:
        printf("Id %s @%d:%d -> ", n->nome, n->linha, n->coluna);
        if (n->escopo) printf("%s [%s, declarado em %d:%d]\n", n->tipo, n->escopo, n->dl, n->dc);
        else printf("não declarado\n");
        break;
    case K_NUM: printf("Num %s @%d:%d\n", n->nome, n->linha, n->coluna); break;
    }
}

/* -- parser --------------------------------------------------------------- */

typedef struct {
    Fluxo *fl;
    char **erros;
    size_t n_erros;
    int blocos;
} Parser;

static void erro_sem(Parser *p, char *msg) {
    p->erros = realoca(p->erros, (p->n_erros + 1) * sizeof(char *));
    p->erros[p->n_erros++] = msg;
}

static const Token *exige(Parser *p, TokenType tipo, const char *esperado) {
    if (fluxo_verifica(p->fl, tipo)) return fluxo_proximo(p->fl);
    erro_sintaxe(fluxo_atual(p->fl), esperado);
    return NULL;
}

static const char *tipo_de(TokenType t) {
    switch (t) {
    case TOK_KW_INT: return "int";
    case TOK_KW_FLOAT: return "float";
    case TOK_KW_CHAR: return "char";
    case TOK_KW_BOOL: return "bool";
    default: return NULL;
    }
}

static No *declara(Parser *p, const Token *t, const char *tipo, Scope *escopo) {
    No *d = no_novo(K_DECL, t->line, t->column, copia(t->lexeme));
    d->tipo = tipo;
    d->escopo = copia(escopo->nome);
    Scope *onde = NULL;
    Simbolo *sombra = escopo->parent ? lookup(escopo->parent, t->lexeme, &onde) : NULL;
    Simbolo *anterior = procura_local(escopo, t->lexeme);
    if (anterior) {
        erro_sem(p, formata("Erro semântico na linha %d, coluna %d: redeclaração de '%s' no escopo %s "
                            "(declarado na linha %d, coluna %d).",
                            t->line, t->column, t->lexeme, escopo->nome, anterior->linha, anterior->coluna));
        d->redecl = 1;
        return d;
    }
    if (sombra) {
        d->sombra = copia(onde->nome);
        d->sl = sombra->linha;
        d->sc = sombra->coluna;
    }
    declare(escopo, t, tipo);
    return d;
}

static No *resolve(Parser *p, const Token *t, Scope *escopo) {
    No *id = no_novo(K_ID, t->line, t->column, copia(t->lexeme));
    Scope *onde = NULL;
    Simbolo *s = lookup(escopo, t->lexeme, &onde);
    if (s) {
        id->tipo = s->tipo;
        id->escopo = copia(onde->nome);
        id->dl = s->linha;
        id->dc = s->coluna;
    } else {
        erro_sem(p, formata("Erro semântico na linha %d, coluna %d: identificador '%s' não declarado.", t->line,
                            t->column, t->lexeme));
    }
    return id;
}

static int parse_item(Parser *p, Scope *escopo, No *destino);

static int parse_decl(Parser *p, Scope *escopo, No *destino) {
    const char *tipo = tipo_de(fluxo_proximo(p->fl)->type);
    const Token *id = exige(p, TOK_IDENT, "identificador");
    if (!id) return ERRO_SINTAXE;
    no_filho(destino, declara(p, id, tipo, escopo));
    while (fluxo_aceita(p->fl, TOK_COMMA)) {
        if (!(id = exige(p, TOK_IDENT, "identificador"))) return ERRO_SINTAXE;
        no_filho(destino, declara(p, id, tipo, escopo));
    }
    return exige(p, TOK_SEMICOLON, "',' ou ';'") ? OK : ERRO_SINTAXE;
}

static int parse_bloco(Parser *p, Scope *pai, No *destino) {
    fluxo_proximo(p->fl); /* '{' (conferido por parse_item) */
    Scope *escopo = enter_scope(pai, formata("bloco%d", ++p->blocos));
    No *bloco = no_novo(K_BLOCK, 0, 0, copia(escopo->nome));
    no_filho(destino, bloco); /* já pendurado: o main libera tudo em caso de erro */
    int codigo = OK;
    while (codigo == OK && !fluxo_verifica(p->fl, TOK_RBRACE)) {
        if (fluxo_verifica(p->fl, TOK_EOF)) {
            erro_sintaxe(fluxo_atual(p->fl), "'}'");
            codigo = ERRO_SINTAXE;
        } else {
            codigo = parse_item(p, escopo, bloco);
        }
    }
    if (codigo == OK) fluxo_proximo(p->fl);
    leave_scope(escopo);
    return codigo;
}

static int parse_uso(Parser *p, Scope *escopo, No *destino) {
    const Token *t = fluxo_proximo(p->fl);
    No *a = no_novo(K_ASSIGN, t->line, t->column, NULL);
    no_filho(destino, a);
    no_filho(a, resolve(p, t, escopo));
    if (!exige(p, TOK_ASSIGN, "'='")) return ERRO_SINTAXE;
    const Token *v = fluxo_atual(p->fl);
    if (v->type == TOK_IDENT) {
        no_filho(a, resolve(p, fluxo_proximo(p->fl), escopo));
    } else if (v->type == TOK_INT) {
        fluxo_proximo(p->fl);
        no_filho(a, no_novo(K_NUM, v->line, v->column, copia(v->lexeme)));
    } else {
        erro_sintaxe(v, "identificador ou número");
        return ERRO_SINTAXE;
    }
    return exige(p, TOK_SEMICOLON, "';'") ? OK : ERRO_SINTAXE;
}

static int parse_item(Parser *p, Scope *escopo, No *destino) {
    const Token *t = fluxo_atual(p->fl);
    if (tipo_de(t->type)) return parse_decl(p, escopo, destino);
    if (t->type == TOK_LBRACE) return parse_bloco(p, escopo, destino);
    if (t->type == TOK_IDENT) return parse_uso(p, escopo, destino);
    erro_sintaxe(t, "declaração, bloco ou atribuição");
    return ERRO_SINTAXE;
}

int main(int argc, char **argv) {
    if (argc != 2) {
        fprintf(stderr, "uso: %s <arquivo | ->\n", argv[0]);
        return ERRO_USO;
    }
    Fluxo fl;
    int codigo = fluxo_abrir(&fl, argv[1]);
    if (codigo != OK) return codigo;

    Parser p = {&fl, NULL, 0, 0};
    Scope *global = enter_scope(NULL, copia("global"));
    No *raiz = no_novo(K_BLOCK, 0, 0, copia("global"));
    while (codigo == OK && !fluxo_verifica(&fl, TOK_EOF)) {
        codigo = parse_item(&p, global, raiz);
    }
    leave_scope(global);
    if (codigo == OK) {
        imprime(raiz, 0);
        for (size_t i = 0; i < p.n_erros; i++) fprintf(stderr, "%s\n", p.erros[i]);
        codigo = p.n_erros ? ERRO_SEMANTICO : OK;
    }
    for (size_t i = 0; i < p.n_erros; i++) free(p.erros[i]);
    free(p.erros);
    no_free(raiz);
    fluxo_fechar(&fl);
    return codigo;
}
